weblate/trans/forms.py

```diff
@@ -904,12 +904,21 @@ def __init__(self, obj, user=None, *args, **kwargs):
         self.obj = obj
 
         # Add components from other projects with enabled shared TM
-        self.components = obj.project.component_set.filter(
-            source_language=obj.source_language
-        ) | Component.objects.filter(
-            source_language_id=obj.source_language_id,
-            project__contribute_shared_tm=True,
-        ).exclude(project=obj.project)
+        if hasattr(obj, "source_language"):
+            self.components = obj.project.component_set.filter(
+                source_language=obj.source_language
+            ) | Component.objects.filter(
+                source_language_id=obj.source_language_id,
+                project__contribute_shared_tm=True,
+            ).exclude(project=obj.project)
+
+        if hasattr(obj, "project") and hasattr(obj.project, "source_language_ids"):
+            self.components = obj.project.component_set.filter(
+                source_language_id__in=obj.project.source_language_ids
+            ) | Component.objects.filter(
+                source_language_id__in=obj.project.source_language_ids,
+                project__contribute_shared_tm=True,
+            ).exclude(project=obj.project)
```

Comment:
This seems weird: a `Component` instance matches both if clauses here, so the first one is effectively never used. Maybe test using `isinstance(obj, Component)` to make the intention more clear?

Reply:
thanks, Good catch. I have added a check for Component in first if clause. Does it make sense to have `if`, `elif` here since in the case when `Component` obj has `project` attr, it will replace `self.components` from the earlier `if` clause. 

I assume, what we want here in case of `project` addon.

1. Filter the components of the `obj`'s project that have a `source_language_id` in `obj.project.source_language_ids`.

2. The combine the queryset with the one that filters all `Component` objects that have a `source_language_id` in `obj.project.source_language_ids` and belong to a project where `contribute_shared_tm` is `True`, excluding the components of `obj`'s project.